README.md
# Quake Timer

A Windows background application to manage cooldown timers for Quake Champions.

## Getting Started

Download the latest `.zip` from [the Releases page](https://github.com/rocheio/quake-timer/releases).

Unzip it, run the `.exe`, and click through the Windows security warnings (fixing the app so it doesn't have these warnings will be done in the future).

In Quake Champions, set your Video Display Mode to `Borderless` or `Windowed`.

Ensure your volume is loud enough to hear the alerts.

Press the registered hotkey at the appropriate time during gameplay:

- Mega Health: `Alt + 1` (30 second respawn)
    - Press immediately after someone picks it up
- Heavy Armor: `Alt + 2` (30 second respawn)
    - Press immediately after someone picks it up
- Quad Damage: `Alt + 3` (120 second respawn)
    - Press immediately after someone picks up the Protection
- Protection: `Alt + 4` (120 second respawn)
    - Press immediately after someone picks up the Quad Damage

The item's name will be announced 10 seconds and 20 seconds before the timer expires.

To close the app, hit `Alt + Ctrl + X` or send a keyboard interrupt in the console (`Ctrl + C`).

## Development

Build and run the app on Windows:

```sh
go build ./cmd/quaketimer
./quaketimer.exe
```

Other operating systems are not yet supported.

## Credits

- Bootstrapping of basic global hotkey listener: https://stackoverflow.com/a/38954281
- Audio files: http://www.fromtexttospeech.com/


CHANGELOG.md
# Changelog

## 0.2.1 - Unreleased
- Add confirmation blip on timer set
- Change alert times to 20 and 10 seconds instead of 30 and 10 seconds

## 0.2.0 - 2018-06-10
- Add `Manager.AddKey` method for more intuitive key creation
- Add `Hotkey.Name` attribute for better reference in output
- Use `time.After` in `Listen` to sleep safely
- Add `KeyPress` logic to parse actual time of press from `WindowsMessage`
- Add `Cooldown` class for more consistent definition

## 0.1.0 - 2018-06-08
- Working prototype with five-second alerts for each major map spawn


pkg/hotkey/hotkey.go
package hotkey

import (
	"bytes"
	"fmt"
	"log"
	"syscall"
	"time"
	"unsafe"
)

const (
	ModAlt = 1 << iota
	ModCtrl
	ModShift
	ModWin
)

type Manager struct {
	User32  *syscall.DLL
	regkey  *syscall.Proc
	peekmsg *syscall.Proc
	// keys a map of hotkeys that are currently registered to the system
	keys map[int16]*Hotkey
	// exit is a flag that can be set from Goroutines to stop Listen
	exit bool
	// two attrs to track the first known Windows time registered
	// and the conversion of that int32 to the Unix time
	firstDur  int32
	firstTime time.Time
	// optional action to take each time a key press is registered
	OnKeyPress func()
}

func NewManager() (*Manager, error) {
	var err error
	m := Manager{
		keys: make(map[int16]*Hotkey),
	}
	m.User32, err = syscall.LoadDLL("user32")
	if err != nil {
		return nil, err
	}
	m.regkey, err = m.User32.FindProc("RegisterHotKey")
	if err != nil {
		return nil, err
	}
	m.peekmsg, err = m.User32.FindProc("PeekMessageW")
	if err != nil {
		return nil, err
	}
	return &m, nil
}

func (m *Manager) AddKey(name string, modifiers int, keycode int, action func(d time.Time)) {
	i := len(m.keys) + 1
	m.keys[int16(i)] = &Hotkey{
		Id:        i,
		Modifiers: modifiers,
		KeyCode:   keycode,
		Name:      name,
		Action:    action,
	}
}

func (m *Manager) RegisterHotkey(i int16, h *Hotkey) error {
	r1, _, err := m.regkey.Call(
		0, uintptr(h.Id), uintptr(h.Modifiers), uintptr(h.KeyCode),
	)

	if r1 == 1 {
		log.Printf("registered %s", h)
		m.keys[i] = h
		return nil
	}

	if err.Error() == "Hot key is already registered." {
		return nil
	}

	return fmt.Errorf("Failed to register %v, error: %v", h, err)
}

func (m *Manager) RegisterHotkeys() error {
	for _, key := range m.keys {
		err := m.RegisterHotkey(int16(key.Id), key)
		if err != nil {
			return err
		}
	}
	return nil
}

func (m *Manager) SeekKeyPress() (*KeyPress, error) {
	msg := &WindowsMessage{}
	m.peekmsg.Call(uintptr(unsafe.Pointer(msg)), 0, 0, 0, 1)

	if m.firstDur == 0 {
		m.firstDur = msg.DWORD
		m.firstTime = time.Now()
	}

	if msg.WPARAM == 0 {
		return nil, nil
	}

	key, ok := m.keys[msg.WPARAM]
	if !ok {
		return nil, fmt.Errorf("key not registered to manager: %d", msg.WPARAM)
	}

	pressTime := m.firstTime.Add(msg.DurationSince(m.firstDur))
	return &KeyPress{key, pressTime}, nil
}

func (m *Manager) Listen() error {
	for {
		select {
		case <-time.After(time.Millisecond * 100):
			if m.exit {
				log.Println("received signal to exit")
				return nil
			}

			p, err := m.SeekKeyPress()
			if err != nil {
				return err
			}
			if p == nil {
				continue
			}

			if m.OnKeyPress != nil {
				m.OnKeyPress()
			}

			log.Printf("%s pressed at %s", p.key, p.time.Format("15:04:05"))
			if p.key.Action != nil {
				go p.key.Action(p.time)
			}
		}
	}
}

func (m *Manager) Exit() {
	m.exit = true
}

type Hotkey struct {
	Id        int    // Unique id
	Modifiers int    // Mask of modifiers
	KeyCode   int    // Key code, e.g. 'A'
	Name      string // User-defined name for hotkey
	// Action does something based on the time of a keypress
	Action func(t time.Time)
}

func (h *Hotkey) String() string {
	mod := &bytes.Buffer{}
	if h.Modifiers&ModAlt != 0 {
		mod.WriteString("Alt+")
	}
	if h.Modifiers&ModCtrl != 0 {
		mod.WriteString("Ctrl+")
	}
	if h.Modifiers&ModShift != 0 {
		mod.WriteString("Shift+")
	}
	if h.Modifiers&ModWin != 0 {
		mod.WriteString("Win+")
	}
	return fmt.Sprintf("Hotkey[%s, %s%c]", h.Name, mod, h.KeyCode)
}

// WindowsMessage is a message from a Windows thread's queue
// https://msdn.microsoft.com/en-us/library/windows/desktop/ms644958(v=vs.85).aspx
type WindowsMessage struct {
	HWND   uintptr
	UINT   uintptr
	WPARAM int16
	LPARAM int64
	DWORD  int32
	POINT  struct{ X, Y int64 }
}

func (w WindowsMessage) DurationSince(firstDur int32) time.Duration {
	return time.Millisecond * time.Duration(w.DWORD-firstDur)
}

// KeyPress is a validated press of a Hotkey at a given time
type KeyPress struct {
	key  *Hotkey
	time time.Time
}


pkg/cooldown/cooldown.go
package cooldown

import (
	"log"
	"time"

	"github.com/rocheio/quake-timer/pkg/audio"
)

func DoAfter(d time.Duration, f func()) {
	select {
	case <-time.After(d):
		f()
	}
}

type Cooldown struct {
	Name      string
	Duration  time.Duration
	AudioFile string
}

func (c Cooldown) Start(t time.Time) {
	remaining := c.Duration - time.Now().Sub(t)
	if remaining > 20*time.Second {
		go DoAfter(remaining-20*time.Second, func() {
			log.Printf("%s in twenty seconds", c.Name)
			err := audio.PlayFiles(c.AudioFile, "./audio/in-twenty-seconds.wav")
			if err != nil {
				log.Fatal(err)
			}
		})
	}
	if remaining > 10*time.Second {
		go DoAfter(remaining-10*time.Second, func() {
			log.Printf("%s in ten seconds", c.Name)
			err := audio.PlayFiles(c.AudioFile, "./audio/in-ten-seconds.wav")
			if err != nil {
				log.Fatal(err)
			}
		})
	}
}


cmd/quaketimer/main.go
package main

import (
	"log"
	"time"

	"github.com/rocheio/quake-timer/pkg/audio"
	"github.com/rocheio/quake-timer/pkg/cooldown"
	"github.com/rocheio/quake-timer/pkg/hotkey"
)

func main() {
	m, err := hotkey.NewManager()
	if err != nil {
		log.Fatal(err)
	}
	defer m.User32.Release()

	m.OnKeyPress = func() {
		audio.PlayFile("./audio/blip.wav")
	}

	m.AddKey("Exit", hotkey.ModAlt+hotkey.ModCtrl, 'X', func(t time.Time) {
		m.Exit()
	})

	cd := cooldown.Cooldown{"Mega Health", 30 * time.Second, "./audio/mega-health.wav"}
	m.AddKey("Mega Health", hotkey.ModAlt, '1', cd.Start)

	cd = cooldown.Cooldown{"Heavy Armor", 30 * time.Second, "./audio/heavy-armor.wav"}
	m.AddKey("Heavy Armor", hotkey.ModAlt, '2', cd.Start)

	cd = cooldown.Cooldown{"Quad Damage", 120 * time.Second, "./audio/quad-damage.wav"}
	m.AddKey("Quad Damage", hotkey.ModAlt, '3', cd.Start)

	cd = cooldown.Cooldown{"Protection", 120 * time.Second, "./audio/protection.wav"}
	m.AddKey("Protection", hotkey.ModAlt, '4', cd.Start)

	err = m.RegisterHotkeys()
	if err != nil {
		log.Fatal(err)
	}

	log.Println("listening for key presses... press Alt+Ctrl+X to exit")
	err = m.Listen()
	if err != nil {
		log.Fatal(err)
	}

	log.Println("program completed")
}
